compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,426 @@
+static bool needsFixup(Type* t);
+static bool validateFormalIntent(FnSymbol* fn, ArgSymbol* as);
+static bool validateReturnIntent(FnSymbol* fn);
+static FnSymbol* createWrapper(FnSymbol* fn);
+static Type* getTypeForFixup(Type* t, bool ret=false);
+static const char* getConversionCallName(Type* srctype, Type* dsttype);
+static FnSymbol* getConversionCall(Type* srctype, Type* dsttype);
+static FnSymbol* resolveConversionCall(Type* srctype, Type* dsttype);
+static VarSymbol* fixupFormal(FnSymbol* wrapper, int idx);
+static void changeRetType(FnSymbol* wrapper);
+static void insertUnwrappedCall(FnSymbol* wrapper, FnSymbol* fn,
+                                const std::vector<VarSymbol*>& tmps);
+
+//
+// Some type information is lost during this pass (IE, converting `string`
+// formals to `c_string`, and this information might be useful when
+// generating code in later passes.
+//
+FnSymbol* getUnwrappedFunction(FnSymbol* wrapper) {
+  std::map<FnSymbol*, FnSymbol*>::iterator it;
+  if ((it = wrapperMap.find(wrapper)) != wrapperMap.end()) {
+    return it->second;
+  }
+  return NULL;
+}
+
+void fixupExportedFunctions(const std::vector<FnSymbol*>& fns) {
+  for_vector(FnSymbol, fn, fns) {
+    fixupExportedFunction(fn);
+  }
+}
+
+void fixupExportedFunction(FnSymbol* fn) {
+  SET_LINENO(fn);
+  attemptFixups(fn);
+}
+
+Type* getCharPtrType(void) {
+  static Type* result = NULL;
+
+  if (result != NULL) { return result; }
+
+  const char* modname = "ExternalString";
+  const char* symname = "chpl__exportTypeCharPtr";
+  ModuleSymbol* esm = NULL;
+  Symbol* alias = NULL;
+
+  // Get a handle to the ExternalString module.
+  forv_Vec(ModuleSymbol, md, allModules) {
+    if (md->modTag == MOD_INTERNAL && !strcmp(md->name, modname)) {
+      esm = md;
+      break;
+    }
+  }
+
+  INT_ASSERT(esm != NULL);
+
+  // Get a handle to the c_ptr(c_char) type from our global alias.
+  for_alist(expr, esm->block->body) {
+    DefExpr* de = toDefExpr(expr);
+    if (de) {
+      Symbol* sym = de->sym;
+      if (!strcmp(sym->name, symname)) {
+        alias = sym;
+        break;
+      }
+    }
+  }
+
+  SymExpr* rctx = new SymExpr(alias);
+
+  // Insert into tree and resolve, just in case.
+  chpl_gen_main->insertAtHead(rctx);
+  resolveExpr(rctx);
+  rctx->remove();
+
+  INT_ASSERT(alias->type != dtUnknown);
+  result = alias->type;
+
+  return result;
+}
+
+static void attemptFixups(FnSymbol* fn) {
+  std::vector<VarSymbol*> tmps;
+  FnSymbol* wrapper = NULL;
+
+  if (!fn->hasFlag(FLAG_EXPORT)) { return; }
+
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+
+    if (needsFixup(as->type) && validateFormalIntent(fn, as)) {
+      if (wrapper == NULL) { wrapper = createWrapper(fn); }
+      VarSymbol* tmp = fixupFormal(wrapper, i);
+      INT_ASSERT(tmp != NULL);
+      tmps.push_back(tmp);
+    } else {
+      // Push back a NULL sentry value for unconverted formals.
+      tmps.push_back(NULL);
```

Comment:
I played with this, and while it would reduce the space cost to O(changed formals) it complicates the loop logic when constructing the new call a bit more than I want.